File: frontend/src/types/activity.ts
export enum ActivityStatus {
  CREATED = 'CREATED',
  DATA_READY = 'DATA_READY',
  SIMILARITY_READY = 'SIMILARITY_READY',
  PREDICTION_READY = 'PREDICTION_READY',
  NOT_PROCESSABLE = 'NOT_PROCESSABLE',
}

export interface Activity {
  id: number;
  name: string;
  distance: number; //meters
  moving_time: number; // seconds
  total_elevation_gain: number; // meters
  athlete_id: number;
  type: string;
  sport_type: string;
  start_date: string; // ISO string
  map_polyline: string;
  average_speed: number; // m/s
  max_speed: number; // m/s
  average_cadence: number | null;
  average_temp: number | null;
  average_watts: number | null;
  weighted_average_watts: number | null;
  has_heartrate: boolean;
  status: ActivityStatus | null;
}

export interface ActivityStream {
  latlng: [number, number][];
  altitude: number[];
  heartrate: number[];
  distance: number[];
  time: number[];
  speed: number[];
  grade: number[];
  cadence: number[];
}

export interface Segment {
  segmentId: string;
  activityId: number;
  startDistance: number;
  endDistance: number;
  avgGradient: number;
  avgCadence: number;
  startLat: number;
  endLat: number;
  startLng: number;
  endLng: number;
  startAltitude: number;
  endAltitude: number;
  startTime: number;
  endTime: number;
  avgSpeed: number;
  elevationGain: number;
  efficiencyScore: number;
  startHeartrate: number;
  endHeartrate: number;
  avgHeartrate: number;
  roadType: string;
  surfaceType: string;
  temperature: number;
  humidity: number;
  wind: number;
  weatherId: number;
  weatherMain: string;
  weatherDescription: string;
  weatherIcon: string;
  efficiencyZone: string;
  gradeEfficiencyZone: string;
  gradeCategory: number;
}

export interface PlannedActivityInput {
  name: string;
  distance: number;
  planned_duration: number; // seconds
  total_elevation_gain: number;
  type: string; // e.g., 'planned'
  sport_type: string; // 'Run' | 'Trail Run'
  start_date: string; // ISO string
}

File: backend/activity-service/src/main/java/com/trailmetrics/activities/model/Activity.java
package com.trailmetrics.activities.model;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import java.time.Instant;
import java.util.List;
import lombok.Data;

@Entity
@Table(name = "activities")
@Data
public class Activity {

  @Id
  private Long id;

  @Column(name = "name")
  private String name;

  @Column(name = "distance")
  private double distance;

  @Column(name = "moving_time")
  private int movingTime;

  @Column(name = "total_elevation_gain")
  private double totalElevationGain;

  @Column(name = "athlete_id")
  private Long athleteId;

  @Column(name = "type")
  private String type;

  @Column(name = "sport_type")
  private String sportType;

  @Column(name = "start_date")
  private Instant startDate;

  @Column(name = "description", columnDefinition = "TEXT")
  private String description;

  @Column(name = "map_polyline", columnDefinition = "TEXT")
  private String mapPolyline;

  @Column(name = "average_speed")
  private double averageSpeed;

  @Column(name = "max_speed")
  private double maxSpeed;

  @Column(name = "average_cadence")
  private Double averageCadence;

  @Column(name = "average_temp")
  private Integer averageTemp;

  @Column(name = "average_watts")
  private Double averageWatts;

  @Column(name = "weighted_average_watts")
  private Double weightedAverageWatts;

  @Column(name = "has_heartrate")
  private Boolean hasHeartrate;

  @OneToMany(mappedBy = "activity", cascade = CascadeType.ALL, orphanRemoval = true)
  private List<ActivityStream> streams;

  @OneToOne
  @JoinColumn(name = "id")
  private ActivityStatusTracker statusTracker;

  @Transient
  private ActivityStatus status;

}

